### PR TITLE
Check transactions in generated xml is as expected

### PR DESCRIPTION
**Pull request by LaurenQurashi**
# TP2000-721 Check transactions in generated xml is as expected

## Why
A Little while ago, we had an issue where some of the footnote associations weren't copied into the envelope. They were in the workbasket, and they got published, they just simply didn't copy down when the command was run. We can't replicate the error at all, so we're not entirely sure what went wrong.. but we can prevent the envelope from being made should it happen again. 

## What
In this PR, I've added a validate_envelope function in the publsihing subsystem that runs the envelopes through some sense checks. It checks that the count of the transactions & tracked models in an envelope match the workbasket it was copied from. 
This fucntion is copied from exporter/serializers that only checked the record order. The fucntion raises a TaricDataAssertionError if it fails a check with a message specific to which check failed.
This version of the validate_envelope check is run in dump_transaction and upload_envelope in the envelope model.

Also included is a check which throws MultipleEnvelopesGenerated error if a workbasket generates more than one envelope.
This hasn't happened but if it does this check will catch it and throw the error, failing envelope generation.

A check on the Envelope questset for deleted xml_file's has been fixed

Ticket created to display error in UI and send email notification
https://uktrade.atlassian.net/browse/TP2000-786

### CODE REVIEW

File: publishing/util.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+# extract the workbaskets check into the function that loops over the list of envelopes
+# and compare the workbaskets expected results with the multiple envelope results returned.
+# This will enable support for multiple envelopes generated from QUEUED workbasketd
+# Currently this workbasket check only checks against a single envelope
+# and would not support the rendering of multiple envelopes
+
+
+def validate_envelope(envelope_file, workbaskets, skip_declaration=False):
+    """
+    Validate envelope content for XML issues and data missing & order issues.
```
Comment (paulpepper-trade) [inline publishing/util.py:86]: Maybe also worth remarking on what `workbaskets` and `skip_declaration` params are.

File: settings/test.py [hunk 2]
```diff
@@ -33,3 +33,7 @@
 
 SKIP_WORKBASKET_VALIDATION = is_truthy(os.getenv("SKIP_WORKBASKET_VALIDATION", True))
 USE_IMPORTER_CACHE = is_truthy(os.getenv("USE_IMPORTER_CACHE", False))
+
+S3_ACCESS_KEY_ID = "test_local_id"
+S3_SECRET_ACCESS_KEY = "test_local_key"
+S3_ENDPOINT_URL = "https://test-s3-url.local/"
```
Comment (paulpepper-trade) [inline settings/test.py:39]: Are these duplicates of the default values in `settings/common.py`?

**Review by paulpepper-trade — CHANGES_REQUESTED** [review]

File: conftest.py [hunk 3]
```diff
@@ -360,6 +364,91 @@ def session_workbasket(client, new_workbasket):
...
+        with factories.ApprovedTransactionFactory.create(workbasket=workbasket):
+            factories.FootnoteTypeFactory()
+            factories.AdditionalCodeFactory()
+        return workbasket
+
+    return factory
+
+
+@pytest.fixture
+def packaged_workbasket_factory(queued_workbasket_factory):
```
[the remark this answers is not in the document]
Comment (a-gleeson) [reply conftest.py:382]: Ah I will try clear that up so currently there isn't much use for the `queued_workbasket_factory()` but it allows for multiple instances of workbaskets to be created in tests.
Here though I have split off creating the workbasket in a queued state.

File: conftest.py [hunk 4]
```diff
@@ -360,6 +364,91 @@ def session_workbasket(client, new_workbasket):
...
+                workbasket=workbasket, **kwargs
+            )
+        return packaged_workbasket
+
+    return factory
+
+
+@pytest.fixture
+def envelope_factory(packaged_workbasket_factory, envelope_storage):
+    def factory(**kwargs):
```
[the remark this answers is not in the document]
Comment (a-gleeson) [reply conftest.py:404]: I'll do that

File: conftest.py [hunk 5]
```diff
@@ -360,6 +364,91 @@ def session_workbasket(client, new_workbasket):
...
+            factories.FootnoteTypeFactory()
+            factories.AdditionalCodeFactory()
+        return workbasket
+
+    return factory
+
+
+@pytest.fixture
+def packaged_workbasket_factory(queued_workbasket_factory):
+    def factory(**kwargs):
```
[the remark this answers is not in the document]
Comment (a-gleeson) [reply conftest.py:383]: I'll do that

File: publishing/models/envelope.py [hunk 6]
```diff
@@ -254,34 +266,45 @@ def upload_envelope(
...
 
-            rendered_envelope = list(
+            rendered_envelopes = list(
                 serializer.split_render_transactions(transactions),
-            )[0]
+            )
+            if len(rendered_envelopes) > 1:
+                msg = f"Multiple envelopes generated for workbasket:  {workbasket.pk}."
+                logger.error(msg)
+                raise MultipleEnvelopesGenerated(msg)
```
[the remark this answers is not in the document]
Comment (a-gleeson) [reply publishing/models/envelope.py:275]: Thanks!

File: settings/test.py (hunk 2 above)
Comment (a-gleeson) [reply settings/test.py:39]: I was meant to remove them from common.py I'll do that

File: publishing/util.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+from django.conf import settings
+from lxml import etree
+
+from common.util import parse_xml
+from common.xml.namespaces import nsmap
+
+logger = logging.getLogger(__name__)
+
+"""
+TrackedModel types make up the keys of the dictionary
```
Comment (paulpepper-trade) [inline publishing/util.py:13]: I think the docstring should go after the declaration: https://peps.python.org/pep-0258/#attribute-docstrings.

**Review by paulpepper-trade — APPROVED** [review]

**Pull request merged**